wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -735,7 +735,31 @@ public function execute_sqlite_query( string $sql, array $params = array() ): PD
 	 */
 	public function begin_transaction(): void {
 		if ( 0 === $this->transaction_level ) {
-			$this->execute_sqlite_query( 'BEGIN' );
+			/*
+			 * When we're executing a statement that will write to the database,
```

Comment:
I'd add [this excerpt](https://www.sqlite.org/lang_transaction.html):

> Transactions can be DEFERRED, IMMEDIATE, or EXCLUSIVE. The default transaction behavior is DEFERRED.
> 
> DEFERRED means that the transaction does not actually start until the database is first accessed.
> 
> IMMEDIATE causes the database connection to start a new write immediately, without waiting for a write statement. The BEGIN IMMEDIATE might fail with [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy) if another write transaction is already active on another database connection.

It also makes me question why BEGIN IMMEDIATE works – this passage suggests it should not. Is there any way we can add a deterministic unit test?

Reply:
@adamziel I added the docs: https://github.com/Automattic/sqlite-database-integration/pull/51/commits/31c319cd644fcb1320cca35f4a4fbf03b6a36a4a

> It also makes me question why BEGIN IMMEDIATE works – this passage suggests it should not.

I think it works because even though upgrading a read transaction to a write transaction fails immediately when another write is in progress, the `BEGIN IMMEDIATE` statement can honor the `busy_timeout`, as there is no need to restart a full transaction.

>  Is there any way we can add a deterministic unit test?

I can give it a shot. Probably not so easy to do in PHP.

Reply:
Thanks! Maybe starting two connections in a single thread would suffice? If we can avoid synchronizing multiple processes it may be pretty easy

Reply:
Ah, I see what you mean. Would this help? https://github.com/clue/reactphp-sqlite

Reply:
@adamziel I think so, yeah, should we add it for this test?